inttest/cplb/cplb_test.go

```diff
@@ -0,0 +1,159 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keepalived
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/k0sproject/k0s/inttest/common"
+
+	"github.com/stretchr/testify/suite"
+)
+
+type keepalivedSuite struct {
+	common.BootlooseSuite
+}
+
+const haControllerConfig = `
+spec:
+  api:
+    externalAddress: %s
+  network:
+    controlPlaneLoadBalancing:
+      vrrpInstances:
+      - virtualIPs: ["%s/24"]
+        authPass: "123456"
+`
+
+// SetupTest prepares the controller and filesystem, getting it into a consistent
+// state which we can run tests against.
+func (s *keepalivedSuite) TestK0sGetsUp() {
+	ipAddress := s.getLBAddress()
+	ctx := s.Context()
+	var joinToken string
+
+	for idx := 0; idx < s.BootlooseSuite.ControllerCount; idx++ {
+		s.Require().NoError(s.WaitForSSH(s.ControllerNode(idx), 2*time.Minute, 1*time.Second))
+		s.PutFile(s.ControllerNode(idx), "/tmp/k0s.yaml", fmt.Sprintf(haControllerConfig, ipAddress, ipAddress))
+
+		// Note that the token is intentionally empty for the first controller
+		s.Require().NoError(s.InitController(idx, "--config=/tmp/k0s.yaml", "--disable-components=metrics-server", joinToken))
+		s.Require().NoError(s.WaitJoinAPI(s.ControllerNode(idx)))
+
+		// With the primary controller running, create the join token for subsequent controllers.
+		if idx == 0 {
+			token, err := s.GetJoinToken("controller")
+			s.Require().NoError(err)
+			joinToken = token
+		}
+	}
+
+	// Final sanity -- ensure all nodes see each other according to etcd
+	for idx := 0; idx < s.BootlooseSuite.ControllerCount; idx++ {
+		s.Require().Len(s.GetMembers(idx), s.BootlooseSuite.ControllerCount)
+	}
+
+	// Create a worker join token
+	workerJoinToken, err := s.GetJoinToken("worker")
+	s.Require().NoError(err)
+
+	// Start the workers using the join token
+	s.Require().NoError(s.RunWorkersWithToken(workerJoinToken))
+
+	client, err := s.KubeClient(s.ControllerNode(0))
+	s.Require().NoError(err)
+
+	s.Require().NoError(s.WaitForNodeReady(s.WorkerNode(0), client))
+
+	// Verify that all servers have the dummy interface
+	for idx := 0; idx < s.BootlooseSuite.ControllerCount; idx++ {
+		s.checkDummy(ctx, s.ControllerNode(idx), ipAddress)
+	}
+
+	// Verify that only one controller has the VIP in eth0
+	count := 0
+	for idx := 0; idx < s.BootlooseSuite.ControllerCount; idx++ {
+		if s.hasVIP(ctx, s.ControllerNode(idx), ipAddress) {
+			count++
+		}
+	}
+	s.Require().Equal(1, count, "Expected only one controller to have the VIP")
+
```

Comment:
We don't seem to actually test LB functionality and a failover case. Not really sure how to do that though

Reply:
There is no LB in this PR, just the virtual IP.

That will come in the second PR once this is merged. Adding LB is almost the same size so it makes things a bit saner to merge.

This tests that the virtual IP is working in two ways:
1- The node joins the virtual IP, so it has to be working
2- Checks that every controller has the dummy interface
3- Checks that only one controller has the VIP in the real interface.

Reply:
Oh, right. Makes sense.